File: cpp/207-course-schedule-i.cpp
#include <vector>
#include <iostream>
#include <queue>
#include <set>
#include <print>
using namespace std;

class Solution
{
public:
    bool canFinish(int numCourses, vector<vector<int>> &prerequisites)
    {
        // maintain a set of remaining nodes
        set<int> nodes;
        for (int i = 0; i < numCourses; i++)
        {
            nodes.insert(i);
        }

        // build the adjacency list
        vector<vector<int>> graph(numCourses);
        for (int i = 0; i < prerequisites.size(); i++)
        {
            graph[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }

        // build the indegree array
        vector<int> indegree(numCourses, 0);
        for (int i = 0; i < prerequisites.size(); i++)
        {
            indegree[prerequisites[i][0]]++;
        }

        // push all the nodes with indegree 0 into the queue
        queue<int> q;
        for (int i = 0; i < numCourses; i++)
        {
            if (indegree[i] == 0)
            {
                q.push(i);
                nodes.erase(i);
            }
        }

        while (!q.empty())
        {
            int curr = q.front();
            q.pop();
            for (int i = 0; i < graph[curr].size(); i++)
            {
                int adj = graph[curr][i];
                indegree[adj]--;
                if (indegree[adj] == 0)
                {
                    q.push(adj);
                    nodes.erase(adj);
                }
            }
        }

        if (!nodes.empty())
        {
            return false;
        }
        return true;
    }
};

int main()
{
    Solution solution;
    int numCourses = 2;
    vector<vector<int>> prerequisites = {{1, 0}, {0, 1}};
    // vector<vector<int>> prerequisites = {{1, 0}};
    cout << (solution.canFinish(numCourses, prerequisites) ? "YES" : "NO") << endl;
    return 0;
}

File: python3/207-course-schedule-i.py
"""
https://www.geeksforgeeks.org/topological-sorting-indegree-based-solution/
MEMORIZE KAHN'S ALGORITHM FOR TOPOLOGICAL SORT
"""

from typing import List
from collections import defaultdict, deque

def canFinish(numCourses: int, prerequisites: List[List[int]]) -> bool:

    ordering_builder = []

    # dictionary
    D = defaultdict(list)
    for (u,v) in prerequisites:
        D[u].append(v)

    # all nodes in the graph
    remaining_nodes = set(range(numCourses))

    # calculate in-degree of each node (aka number of dependencies)
    in_degree = [0] * numCourses
    for node in range(numCourses):
        for adj_node in D[node]:
            in_degree[adj_node] += 1

    # add all nodes with in-degree of zero to queue
    bfs_queue = deque()
    for node in range(numCourses):
        if in_degree[node] == 0:
            bfs_queue.append(node)
            remaining_nodes.remove(node)

    while bfs_queue:
        # add nodes without dependencies to ordering
        node = bfs_queue.popleft()
        ordering_builder.append(node)


        # for each adjacent node, decrease the number of dependencies by 1
        for adj_node in D[node]:
            in_degree[adj_node] -= 1
            if in_degree[adj_node] == 0:
                bfs_queue.append(adj_node)
                remaining_nodes.remove(adj_node)

    # cycle detection
    if remaining_nodes:
        return False
    return True
    
numCourses = 4
prerequisites = [[1,0],[2,0],[3,1],[3,2]]

print(canFinish(numCourses, prerequisites))